cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java

```diff
@@ -490,6 +516,151 @@ private void closeScanner() {
     }
   }
 
+  /**
+   * Filters elements matching an index  {@link ScannerIterator}.
+   */
+  @VisibleForTesting
+  static final class FilterByIndexIterator extends AbstractCloseableIterator<StructuredRow> {
+    private final Iterator<? extends CloseableIterator<StructuredRow>> scannerIterator;
+    private CloseableIterator<StructuredRow> currentScanner;
+    private final Field<?> filterIndex;
+    private final StructuredTableSchema schema;
+
+    FilterByIndexIterator(Iterator<? extends CloseableIterator<StructuredRow>> scannerIterator, Field<?> filterIndex,
+                          StructuredTableSchema schema) {
+      this.scannerIterator = scannerIterator;
+      this.currentScanner = scannerIterator.hasNext() ? scannerIterator.next() : CloseableIterator.empty();
+      this.filterIndex = filterIndex;
+      this.schema = schema;
+    }
+
+    @Override
+    protected StructuredRow computeNext() {
+      // Postfiltering on scanned rows by index
+      while (!currentScanner.hasNext() && scannerIterator.hasNext()) {
+        closeScanner();
+        currentScanner = scannerIterator.next();
+      }
+      while (currentScanner.hasNext()) {
+        StructuredRow row = currentScanner.next();
+        if (Objects.equals(getFromRow(row), filterIndex.getValue())) {
+          return row;
+        }
+      }
+      return endOfData();
+    }
+
+    @Override
+    public void close() {
+      closeScanner();
+    }
+
+    private Object getFromRow(StructuredRow row) throws InvalidFieldException {
+      switch (filterIndex.getFieldType()) {
```

Comment:
it's better to avoid this comparison for each element, as the type will always be the same for every element in the scan. This can be done by examining the column type in the constructor, then using an appropriate method to get and compare the value.

Reply:
so I am unsure how this can be moved to the constructor, because the method operates on the elements of the iterator `StructuredRow` is obtained in the `computeNext` function

Reply:
I think Albert is saying that the field type info can be retrieved from`filterIndex` already, no need to use `switch` every time you retrieve the element

Reply:
you can do something like:

```

static final class FilterByIndexIterator ... {
  private final Predicate<StructuredRow> predicate;
  ...

  FilterByIndexIterator(...) {
    switch (filterIndex.getFieldType()) {
      case INTEGER:
        predicate = row -> Objects.equals(row.getInteger(filterIndex.getName()), filterIndex.getValue());
      case LONG:
        ...
    }
  }
}

```

Reply:
got it let me try. Thanks!

Reply:
@de-lan so the method itself operates on `row` which is specific to the type of column. So it is not just dependent on `filterIndex`. But the above should work